2要素認証(2FA)に必要なbackend側の実装を追加

rakushoo opened the pull request:
* Resolve #156 

## 変更の概要

* OAuthログインを、/auth/login,signupではなく新規追加した/auth/oauthloginで初回登録と2回目以降のログインを実施するように変更
* 今回のプルリクではbackend側のAPIの追加のみ
 
## なぜこの変更をするのか

* 2FAの実装にあたり、ログインしたユーザが2FAをONしてるか判定するため、DBの内容を見たい場面があった
* 現状の、メール認証と同じAPIからログインする方法だと、一度ログインしてから2FA判定する形となり不自然だったため、他サイトを参考にし、backend側に/auth/oauthlogin のAPIを追加し、ログイン前に確認ができるようにした。

## やったこと

* [x] DBのUserに、oauthid, has2FA, twoFASecret のカラムを追加
* [x] backendのAPIに/auth/oauthloginを追加。初回登録とログイン、usernameの重複を避ける処理などを追加
* [x] frontendのindex.tsxでメールログインと同じ流れをやめ、OAuth用のAPI(/auth/oauthlogin)を叩くように変更
* [x] backendに qr2fa, send2facode, has2fa, validate2fa, disable2fa のAPIを追加

## 変更内容

* UIの変更ならスクリーンショット
* /auth/oauthlogin に、oauthidとimagepathをpostすると、{message:"OK"}が返る

## 影響範囲

* ユーザに影響すること
なし
* メンバーに影響すること
DBのカラム追加があったので、backendでのmigration dev→generateと、frontendのdb pull→generateが必要
* システムに影響すること
なし

## どうやるのか

* 使い方
1. indexから42またはgoogleアイコンを押す
2. 各サービスの認証を通す
3. dashboard画面に遷移する（DBにoauthid,username,avatorPathなどが登録される）
    42アカウントではavatorPathは未対応。

## 課題

* DBに残すuniqueなoauthidにemailを設定している。id(skohrakuなど)を登録するように変更予定
* avatorPathにURLを設定しているためプロフィール画面で表示されない

* 悩んでいること
* とくにレビューしてほしいところ

## 備考

* その他に伝えたいこと

rakushoo committed:
fix: 2FAの準備として、backendにOAuth認証用のAPIを追加

OAuth認証時はoauthidをキーとし、usernameが一意となるように名前をつける処理を追加

rakushoo committed:
add: frontendのschemaのdb pullし忘れ

rakushoo committed:
add: 2FA向けのAPIを追加

追加ライブラリ：qrcode, speakeasy
追加API：
auth/qr2fa/:id: QRコード表示
auth/send2facode: 2FA機能ON時
auth/has2fa: 2FAの有無
auth/validate2fa: ログイン時認証
auth/disable2fa: 2FA機能OFF時

rakushoo committed:
add: DBのUserにtwoAFSecretを追加

rakushoo committed:
fix: 42アカウントからOAuth認証ができない不具合修正

session情報にimageが入っていなかったため、一時的に削除した。
access_tokenを使用してusernameと同時に改めて取得する。

yuta-fujimoto commented:
コンフリクトけっこうしてますね、、
@rakushoo

takashi247 committed:
Merge branch 'main' into add-2factorAuth

takashi247 commented:
> コンフリクトけっこうしてますね、、 @rakushoo

@rakushoo 私の実装部分と関係しているコンフリクトが多かったので、コンフリクト解消しておきました。

takashi247 commented on backend/src/auth/auth.service.ts:
```diff
@@ -90,4 +98,197 @@ export class AuthService {
+      } else {
+        username = dto.oauthid;
+      }
+      //パスワードは
+      const password = Math.random().toString(36).slice(-16);
+      const hashed = await bcrypt.hash(password, 12);
```
OAuthでログインする際にはパスワードは登録しない方がよい、ということで以前合意したと思っていたんですが、ここでパスワードを作成している理由って何かありますか？

rakushoo replied:
参考コードからもらってきていました。
はい、OAuthの時にはパスワード不要ですね、、
schemaの方もString?のため、削除しておきます。

takashi247 commented on backend/src/auth/auth.service.ts:
```diff
@@ -90,4 +98,197 @@ export class AuthService {
+      let username = '';
+      if (sameUsername) {
+        // usernameが使われてたら、一致しないものを生成する
+        while (sameUsername) {
+          username =
+            dto.oauthid + '_' + Math.floor(Math.random() * 100000).toString();
+          sameUsername = await this.prisma.user.findUnique({
+            where: { name: username },
+          });
```
以前まつかわさんが共有してくれていた、日時を末尾につける方法でユーザー名を作成すれば、一発でユニークな名前を作れるのでwhileループにする必要がなくなると思います。

参考: https://github.com/ryo-manba/ft_transcendence/pull/131#discussion_r1036741709

rakushoo replied:
提案ありがとうございます！
Dateを使ってwhileループを止める実装に変更しました。

takashi247 commented on backend/src/auth/auth.service.ts:
```diff
@@ -90,4 +98,197 @@ export class AuthService {
+          },
+        });
+      } catch (error) {
+        if (error instanceof PrismaClientKnownRequestError) {
+          // Prismaが新規作成時に発行するエラー。
+          if (error.code === 'P2002') {
+            throw new ForbiddenException('username is already taken');
+          }
+        }
```
上の部分で、ユーザー名は必ず重複しないように設定しているので、この部分のエラー処理は不要かと思います。

rakushoo replied:
ありがとうございます。
こちらの判定文は削除します。

takashi247 commented on frontend/prisma/schema.prisma:
```diff
@@ -87,6 +87,8 @@ model User {
   point           Int               @default(1000)
   avatarPath      String?
+  has2FA          Boolean           @default(false)
+  oauthid         String?           @unique
```
変数名はcamel caseで統一したいです。（oauthidではなくoAuthId）

rakushoo replied:
承知しました。
oauthid->oAuthId
imagepath->imageParh
に変更します。
（今後も名称気をつけます🙇‍♂️）

takashi247 commented on backend/src/auth/auth.service.ts:
```diff
@@ -90,4 +98,197 @@ export class AuthService {
+  }
+
+  async disable2FA(data: SecretCodeDto): Promise<string> {
+    try {
+      const user_db = await this.prisma.user.findUnique({
+        where: {
+          id: Number(data.userId),
+        },
+      });
+    } catch (error) {
+      console.log(error);
+      throw error;
+    }
+
```
ここの部分は不要かと思います。直接updateしてみて、もしユーザーが見つからなければその時点で例外が投げられると思うので。
参考: https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#update

rakushoo replied:
ありがとうございます。
disable2FA自体の動作確認はこれからなのですが、
直接updateしてもRecordNotFoundのExceptionって事ですね。
こちらのfindUniqueは削除します。

takashi247 commented on backend/src/auth/dto/auth.dto.ts:
```diff
@@ -9,3 +9,11 @@ export class AuthDto {
+export class OauthDto {
+  @IsString()
+  @IsNotEmpty()
+  oauthid: string;
+
+  @IsString()
+  imagepath: string;
```
oAuthId, imagePathにしたいです。

takashi247 commented on backend/src/auth/auth.controller.ts:
```diff
@@ -67,4 +69,54 @@ export class AuthController {
+  }
+
+  @Get('has2fa')
+  has2fa(@Param('id') id: string) {
```
下の関数名と平仄を合わせるのであれば、faは大文字にしたほうがよいかと思います。

takashi247 commented on backend/src/auth/auth.service.ts:
```diff
@@ -90,4 +98,197 @@ export class AuthService {
+    return 'ok';
+  }
+
+  public async has2fa(userId: number): Promise<string> {
```
ここもhas2FAの方がよいかと思います。

takashi247 reviewed (commented):
ありがとうございます。気づきのコメントをいくつかお送りします。

- 42のOAuthでサインアップ、ログインしようとしてもできない
(これはまだGoogleしか対応していないということでしたらご放念ください。)
- avatarPathがちゃんと機能してみない
(ここはこのPRがマージされたら私の方で対応します。イシューも立てておきました。https://github.com/ryo-manba/ft_transcendence/issues/175)
- ログイン画面の42とGoogleのロゴをButtonタグで囲むべき
(今回のPRと直接関係なくて恐縮ですが、現状、ロゴにマウスオーバーしても指マークに変わらないので、ボタンになっていることがわかりにくいかと思います)
- 変数名の命名方法はcamel caseで統一すべき
(oauthidではなくoAuthIdの方が読みやすい。いくつか気づいた部分はコード自体にもコメントつけましたが、その他にももしずれている部分があれば修正してもらえるとありがたいです。)

yuta-fujimoto commented on backend/src/auth/auth.service.ts:
```diff
@@ -90,4 +98,197 @@ export class AuthService {
+      if (!user)
+        throw new HttpException('User not found', HttpStatus.NOT_FOUND);
+
+      const secret = speakeasy.generateSecret();
```
secretってどのタイミングでDBに保存するんですか？`generateQrCode`を呼び出すたびに更新する感じですか？

場合によってはこの処理はフロントエンド側でやってもよさそうだなと思ったので

rakushoo replied:
今回はAPIの追加だけだったため、まだ認証動作の実装は未完成でした。すみません。

回答としては、
generateQrCode（ユーザーが2要素認証をONにする、を選択した場合に実行）を呼び出す度にsecretを生成します。同時に、それに対応したQRコードを表示します。
secretの生成とDBへの登録、画像の作成をbackendで実施し、frontendへはQRコードの画像だけ送ります。
secretの情報はbackend内で閉じた方が良いと考えているので、このようにしています。
（参考コードの動作から、そう理解しました）


generateQrCode　→ QRコードを表示。この時のSecretをDBに保存する、has2FA=falseのまま
send2FACode → ユーザーがアプリで取得した認証コードを入力し、登録する。DBのSecretとのverifyが通れば、has2FA=trueに変更。
という流れです。

yuta-fujimoto replied:
ありがとうございます、qrcode生成のたびにdbに保存する形なのでバックエンド側でやるべきですね

yuta-fujimoto commented on backend/src/auth/auth.service.ts:
```diff
@@ -90,4 +98,197 @@ export class AuthService {
+      throw new Error('hoge');
+    }
+
+    return 'loginと同じくユーザー情報を返したい';
```
https://github.com/ryo-manba/ft_transcendence/pull/153#discussion_r1040987011

まつかわさん曰くfindUniqueは例外出さないらしいので`user_db`返したらだめですか？

rakushoo replied:
ありがとうございます。
はい、validateが成功したらDBのuser情報を返すようにしました。
まだfrontendと繋げていないので動作未確認ですので、次回PRで確認予定です🙏

yuta-fujimoto replied:
ありがとうございます、確認しました

rakushoo committed:
Merge branch 'add-2factorAuth' of https://github.com/ryo-manba/ft_transcendence into add-2factorAuth

rakushoo committed:
fix: OAuthのログインではパスワード不要のため削除

rakushoo committed:
fix: 重複するusernameの場合、Dateを利用して重複しないものを1回作成するように変更

rakushoo committed:
fix: 重複しないIDを生成する事でP2002のエラーは不要になるため、判定処理を削除

rakushoo committed:
update: DBカラム名をoauthidからoAuthIdに変更

併せてdtoもimagePathに変更

rakushoo committed:
fix: 不要なfinduniqueの処理を削除

後のupdateを実行しても同様に例外が発生するため、不要なコードを削除した。

rakushoo committed:
update: 関数名称を変更

rakushoo committed:
update: 2FAのAPI実施動作を修正

QRコード生成時にSecretをDBに登録
ユーザーの初回登録を受けて、2FAをtrueにするようにした。
ログイン時の2FA認証が通った際はユーザ情報をJwtで返すように変更。

rakushoo commented:
レビューありがとうございました！
@takashi247 コンフリクトの解消も助かりました！
反応遅れて申し訳ないです。
ひとまず今回のPRはbackendのAPI追加を中心に実装しています。
問題なければマージお願いします🙏

ブランチは削除しないでもらえると助かります。
→削除して別名で作り直します。

ryo-manba commented on backend/.gitignore:
```diff
@@ -32,4 +32,7 @@ lerna-debug.log*
+!.vscode/extensions.json
+
+# local env files
+.env*.local
```
backendにenv.localを保存することってあるんでしたっけ？

rakushoo replied:
不要です！
削除しておきます。

ryo-manba commented on backend/prisma/schema.prisma:
```diff
@@ -24,6 +24,9 @@ model User {
   followers       FriendRelation[]  @relation("follower")
   followings      FriendRelation[]  @relation("following")
+  oAuthId         String?           @unique
+  has2FA          Boolean           @default(false)
+  twoFASecret     String            @default("")
```
- twoFASecret
- has2FA

'2' か 'two' のどっちかに統一したいですね。

rakushoo replied:
twoFASecret -> secret2FA にリネームしますね。

ryo-manba commented on backend/src/auth/auth.controller.ts:
```diff
@@ -67,4 +69,54 @@ export class AuthController {
   }
+
+  @HttpCode(HttpStatus.OK)
+  @Post('oauthlogin')
```
>   @Post('oauthlogin')

2つの単語が混ざってるので、 `oauth-login` の方がいいかもです。
または `/login/oauth` にして、 `/login` の場合は通常のログインを実行するみたいな感じでもありだと思います。

以下参考になるかと。
https://qiita.com/NagaokaKenichi/items/6298eb8960570c7ad2e9#4%E5%8D%98%E8%AA%9E%E3%82%92%E3%81%A4%E3%81%AA%E3%81%92%E3%82%8B%E5%BF%85%E8%A6%81%E3%81%8C%E3%81%82%E3%82%8B%E5%A0%B4%E5%90%88%E3%81%AF%E3%83%8F%E3%82%A4%E3%83%95%E3%83%B3%E3%82%92%E5%88%A9%E7%94%A8%E3%81%99%E3%82%8B

rakushoo replied:
ありがとうございます！
oauthloginは、OAuthの初回singupとloginと兼ねてるので、/login/oauthよりoauth-loginにしたいと思います。

ryo-manba commented on backend/src/auth/auth.controller.ts:
```diff
@@ -67,4 +69,54 @@ export class AuthController {
+    return this.authService.send2FACode(Number(id), dto);
+  }
+
+  @Get('has2fa')
+  has2FA(@Param('id') id: string) {
+    return this.authService.has2FA(Number(id));
+  }
+
+  @Post('validate2fa')
+  validate2FA(@Body() data: SecretCodeDto) {
+    return this.authService.validate2FA(data);
+  }
+
+  @Post('disable2fa')
+  disable2FA(@Body() data: SecretCodeDto) {
+    return this.authService.disable2FA(data);
+  }
```
なにかリターンするのであれば、戻り値の型が書いてあると嬉しいです！

rakushoo replied:
承知しました！
controllerは戻り値書かなくても通るけどわかりやすく、という事ですね。
今後追加するようにします。

ryo-manba commented on backend/src/auth/auth.service.ts:
```diff
@@ -90,4 +98,174 @@ export class AuthService {
+    });
+
+    if (!user) {
+      // 初めてOAuth認証した時はユーザー登録する
+      const sameUsername = await this.prisma.user.findUnique({
+        where: { name: dto.oAuthId },
+      });
```
whereの条件文のnameに対応するものが oAuthIdで正しいですか？

rakushoo replied:
はい、これは意図通りです。

初回OAuthでのユーザー登録時には、oAuthIdをnameに登録するので、
nameとoAuthIdが一致するものを探して既にその名前が使われている場合には、
name=oAuthId+uniqueSuffix の名前を作成して、登録を行います。

ryo-manba commented on backend/src/auth/auth.service.ts:
```diff
@@ -90,4 +98,174 @@ export class AuthService {
+            avatarPath: dto.imagePath,
+          },
+        });
+      } catch (error) {
+        throw error;
+      }
```
catchして何もせずにそのままthrowするのであれば、この処理不要かと思います！
呼び出し元でcatchするだけでいいです！

ryo-manba replied:
なぜtrycatchを使用するのかを一度考えていただけると、いいかもしれません。
throwも使用しすぎると、補足漏れが起きたりするので注意が必要ですね。

rakushoo replied:
こちらも削除しました。

ryo-manba commented on backend/src/auth/auth.service.ts:
```diff
@@ -90,4 +98,174 @@ export class AuthService {
+          oAuthId: dto.oAuthId,
+        },
+      });
+      if (!user) throw new ForbiddenException('username or password incorrect');
```
これってユーザーの新規作成に成功したかを確かめているのであれば、不要かと思います！
`this.prisma.user.create` が成功すれば保存できてるので。

rakushoo replied:
ここは、その後の処理でユーザーがhas2FAフラグ立っているか、確認するためにuserを取得していました。
ご指摘の通り、createに成功してるので、その後のfindUniqueでも成功するはずなので、140行目の判定は不要と考え、削除しておきます。

ryo-manba commented on backend/src/auth/auth.service.ts:
```diff
@@ -90,4 +98,174 @@ export class AuthService {
+
+      return qr_code;
+    } catch (error) {
+      throw error;
```
ここもthrowするだけならtry catchを使わなくていいですね

rakushoo replied:
確かにそうですね、Cだとreturnしたものをどこで受け取るか考えるのですが、
とりあえず実装、の状態になってました。すみません。
同じく、throwするだけのものは削除しました。

ryo-manba commented on backend/src/auth/auth.service.ts:
```diff
@@ -90,4 +98,174 @@ export class AuthService {
+    return 'ok';
+  }
+
+  public async has2FA(userId: number): Promise<string> {
```
`public async` にしているのってどういう意味があるのでしょうか？

rakushoo replied:
参考コードからそのまま持ってきていました。
現状はpublicを記述する必要性はないので削除しておきます。

ryo-manba commented on backend/prisma/migrations/20221203150735_add_oauth_to_user/migration.sql:
```diff
@@ -0,0 +1,12 @@
+/*
```
マイグレーションファイルは最終決定したもの以外は不要なので、一つだけpushしていただきたいです！

例えば、
- 何かのイベントで限定的にテーブルを追加したい
- イベント終了時にテーブルを元の状態に戻したい

となったときに、migrationファイルを1つ削除するだけで元の状態に戻せます。

上記理由から、テーブル定義を更新する場合はそれに対応したマイグレーションファイルは1つだけが望ましいです。

以下はLaravelの例ですが、基本的に考え方は同じです。

> マイグレーションファイルはデータベースの設計書のようなもので、設計書を記述し終わった後で Artisan migrate コマンドを実行すると、Laravel がその内容をもとに自動でデータベースを構成します。 SQL が書けなくてもデータベースを作れるのは、とても便利ですね。2022/03/25

https://knowledge.cpi.ad.jp/howto-cpi/laravel-migration/#:~:text=%E3%83%9E%E3%82%A4%E3%82%B0%E3%83%AC%E3%83%BC%E3%82%B7%E3%83%A7%E3%83%B3%E3%83%95%E3%82%A1%E3%82%A4%E3%83%AB%E3%81%AF%E3%83%87%E3%83%BC%E3%82%BF%E3%83%99%E3%83%BC%E3%82%B9%E3%81%AE,%E3%81%AF%E3%80%81%E3%81%A8%E3%81%A6%E3%82%82%E4%BE%BF%E5%88%A9%E3%81%A7%E3%81%99%E3%81%AD%E3%80%82

rakushoo replied:
承知しました。
指摘のあった名称変更など、追加してしまったmigrationfileは削除し、
最終盤の名前で新規に作成したmigrationのみを追加しました。

rakushoo committed:
fix: backendでは.env.localは不要なため削除

rakushoo committed:
update: DBのカラム名をtwoFASecretからsecret2FAに変更

rakushoo committed:
update: OAuthのログインAPI名称を変更

rakushoo committed:
update: auto.contorllerに戻り値の定義を追加。

rakushoo committed:
fix: OAuthでの初回登録時の不要なtry,catchを削除。

rakushoo committed:
fix: try,catchでthrowするだけの処理を削除。

rakushoo committed:
fix: 必要か不明だったpublic記述を削除

rakushoo committed:
fix: OAuthに関連するDBのmigrationを一つにまとめた

takashi247 reviewed (approved):
諸々対応ありがとうございました！私のコメントさせて頂いた部分の修正は確認できましたので、Approveします。

ryo-manba commented on backend/src/auth/dto/auth.dto.ts:
```diff
@@ -9,3 +9,11 @@ export class AuthDto {
   @IsNotEmpty()
   username: string;
 }
+export class OauthDto {
+  @IsString()
+  @IsNotEmpty()
+  oAuthId: string;
+
+  @IsString()
+  imagePath: string;
+}
```
`create-oauth.dto.ts` に `CreateOAuthDto` を作るといいかも:hatched_chick:

rakushoo replied:
別ファイルcreate-oauth.dto.tsを作成しそちらで定義しました。

rakushoo committed:
update: dtoの名称を変更

rakushoo committed:
update: OAuthDtoの名称を変更

rakushoo committed:
delete: old migration

ryo-manba commented on backend/prisma/migrations/20221211082253_add_user_colomun_for_2fa/migration.sql:
```diff
@@ -0,0 +1,17 @@
+  - A unique constraint covering the columns `[oAuthId]` on the table `User` will be added. If there are existing duplicate values, this will fail.
+
+*/
+-- DropIndex
+DROP INDEX "User_oauthid_key";
```
この処理が不要なため以下のエラーが発生してしまいます。
```
Error: P3006

Migration `20221211082253_add_user_colomun_for_2fa` failed to apply cleanly to the shadow database. 
Error:
db error: ERROR: index "User_oauthid_key" does not exist
   0: sql_migration_connector::validate_migrations
             at migration-engine/connectors/sql-migration-connector/src/lib.rs:283
   1: migration_core::state::DevDiagnostic
             at migration-engine/core/src/state.rs:251n
```

ryo-manba replied:
以下の実行をお願いします。
- `backend/prisma/migrations/20221211082253_add_user_colomun_for_2fa/migration.sql` を削除する
- make する

rakushoo replied:
大変失礼しました。上記の残っていたmigrationを削除し、
再度、prisma migrate したファイルを追加しました。

yuta-fujimoto reviewed (approved):
遅れましたが僕のほうは大丈夫なのでまつかわさんのほうの修正をお願いします！！

rakushoo committed:
fix: migrationファイルの再作成

ryo-manba reviewed (approved):
確認できました。ありがとうございます！
(makeを実行すると、frontendのschema.prismaに差分がでますが、別のタイミングでフォーマットを適用したもの上げていただければと思います。)
